src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs

```diff
@@ -274,38 +274,50 @@ ExpressionSyntax initializer
 		) {
 			List<ExpressionSyntax> assignments = new() {initializer};
 
-			// If the containing symbol isn't a named type (class) then there
-			// will be no constructors
+			// If the containing symbol isn't a named type,
+			// then there is no .Constructors member
 			if( symbol.ContainingSymbol is not INamedTypeSymbol namedTypeSymbol ) {
 				return assignments.ToImmutableArray();
 			}
 
-			// Retrieve syntax of all constructors
-			foreach(
-				var constructorSyntax in
-				from syn in
-					from sym in namedTypeSymbol.Constructors
-					where sym.DeclaringSyntaxReferences.Length != 0
-					select sym.DeclaringSyntaxReferences
-					          .Single()
-					          .GetSyntax() as ConstructorDeclarationSyntax
-				where syn != null
-				select syn
-			) {
-				var semanticModel = m_compilation.GetSemanticModel( constructorSyntax.SyntaxTree );
-
-				// Add all assignments of which the left side is the
-				// field or property in question
-				assignments.AddRange(
-					from assignment in constructorSyntax.DescendantNodes()
-					                                    .OfType<AssignmentExpressionSyntax>()
-					let compareSymbol = semanticModel.GetSymbolInfo( assignment.Left )
-					                                 .Symbol
-					where SymbolEqualityComparer.Default.Equals( symbol, compareSymbol )
-					select assignment.Right
-				);
+			// If the symbol somehow has no syntax, ignore it
+			if( symbol.DeclaringSyntaxReferences.IsEmpty ) {
+				return assignments.ToImmutableArray();
 			}
 
+			// Retrieve the semantic model so that we can get a symbol
+			// from some syntax later
+			var semanticModel = m_compilation.GetSemanticModel(
+				symbol.DeclaringSyntaxReferences.Single().GetSyntax().SyntaxTree
+			);
+
+			// Retrieve a list of assignment expressions from any constructors
+			// within the class
+			var assignmentExpressions = namedTypeSymbol
+			                           .Constructors
+			                           .Where( sym => !sym.IsImplicitlyDeclared )
+			                           .Select( sym => sym.DeclaringSyntaxReferences
+			                                              .Single()
+			                                              .GetSyntax() )
+			                           .SelectMany( syn => syn.DescendantNodes() )
+			                           .OfType<AssignmentExpressionSyntax>();

```

Comment:
Ok, this looks good, a few things:

* nit: the right alignment of the `.` here is wrong. Is this resharper, or did you configure your editor to do this? Your editor should [be doing this](https://github.com/Brightspace/D2L.CodeStyle/blob/master/src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs#L34-L41). It's not a big deal, it's just that if everyone disagrees then every PR is going to constantly fight between styles. Consistency is key.
* It's probably fine to combine the and `Select`+`Where` and `Select` from below on to the end of here, but I would combine the `Select`+`Where` into a single `Select` with a helper, e.g.

```cs
.Where( ae => IsAnAssignmentTo( ae, symbol ) )
.Select( ae => ae.Right );
```

The first `Select` that creates an anonymous type to coordinate stuff between the next `Where` and `Select` makes it hard to read.

* In `IsAnAssignmentTo`, instead of passing the semantic model down, grab it each time based on the passed in assignment expression. What we're doing here I think has a bug, because `partial` classes means that the semantic model for the field/property may not have include the constructor that writes to it (because it could be in a different file/document)
* nit: it's a bit weird to create a single element `List`, compose it into an `IEnumerable` with `AddRange` and then convert it to `ImmutableArray`. You can do `assignments.Append( initializer ).ToImmutableArray()` to achieve the same end result without the extra allocation.